compiler/plugins/target/ROCM/test/opt_pass_plugin/GPUHello.cpp

```diff
@@ -0,0 +1,79 @@
+#include <iostream>
+#include <vector>
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+using namespace llvm;
+using namespace std;
+
+namespace {
+
+struct GpuHello : public PassInfoMixin<GpuHello> {
+  PreservedAnalyses run(Module &module, ModuleAnalysisManager &) {
+    bool modifiedCodeGen = runOnModule(module);
+
+    return (modifiedCodeGen ? llvm::PreservedAnalyses::none()
+                            : llvm::PreservedAnalyses::all());
+  }
+  bool runOnModule(llvm::Module &module);
+  // isRequired being set to true keeps this pass from being skipped
+  // if it has the optnone LLVM attribute
+  static bool isRequired() { return true; }
+};
+
+} // end anonymous namespace
+
+bool GpuHello::runOnModule(llvm::Module &module) {
+  bool modifiedCodeGen = false;
+  for (auto &function : module) {
+    if (function.isIntrinsic())
+      continue;
+    StringRef functionName = function.getName();
+    if (function.getCallingConv() == CallingConv::AMDGPU_KERNEL ||
+        function.getCallingConv() == CallingConv::PTX_Kernel ||
+        functionName.contains("kernel")) {
+      for (Function::iterator basicBlock = function.begin();
+           basicBlock != function.end(); basicBlock++) {
+        for (BasicBlock::iterator inst = basicBlock->begin();
+             inst != basicBlock->end(); inst++) {
+          DILocation *debugLocation =
+              dyn_cast<Instruction>(inst)->getDebugLoc();
+          std::string sourceInfo;
+          if (!debugLocation) {
+            sourceInfo = (function.getName()).str();
+          } else {
+            sourceInfo =
+                (function.getName() + "\t" + debugLocation->getFilename() +
+                 ":" + Twine(debugLocation->getLine()) + ":" +
+                 Twine(debugLocation->getColumn()))
+                    .str();
```

Comment:
This is a bit dense, consider using `llvm::formatv` or writing with raw_string_ostream`. This could also be a helper function to simplify the loop code a bit.